File: java/com/google/copybara/git/gerritapi/GerritApiTransportImpl.java
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git.gerritapi;

import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.HttpHeaders;
import com.google.api.client.http.HttpRequest;
import com.google.api.client.http.HttpRequestFactory;
import com.google.api.client.http.HttpResponse;
import com.google.api.client.http.HttpResponseException;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.http.json.JsonHttpContent;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.JsonObjectParser;
import com.google.api.client.json.gson.GsonFactory;
import com.google.common.base.Preconditions;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.git.GitCredential.UserPassword;
import com.google.copybara.git.GitRepository;

import java.io.IOException;
import java.lang.reflect.Type;
import java.net.URI;
import java.time.Duration;

import javax.annotation.Nullable;

/**
 * Implementation of {@link GerritApiTransport} that uses direct http calls.
 */
@SuppressWarnings("TypeParameterUnusedInFormals")
public class GerritApiTransportImpl implements GerritApiTransport {

  private static final JsonFactory JSON_FACTORY = new GsonFactory();

  private final GitRepository repo;
  private final URI uri;
  private final HttpTransport httpTransport;

  public GerritApiTransportImpl(GitRepository repo, URI uri, HttpTransport httpTransport) {
    this.repo = repo;
    this.uri = Preconditions.checkNotNull(uri);
    this.httpTransport = Preconditions.checkNotNull(httpTransport);
  }

  @Override
  public <T> T get(String path, Type responseType)
      throws RepoException, ValidationException {
    HttpRequestFactory requestFactory = getHttpRequestFactory(
        getCredentialsIfPresent(uri.toString()));
    GenericUrl url = getUrl(path);
    try {
      return execute(responseType, requestFactory.buildGetRequest(url));
    } catch (IOException e) {
      throw new RepoException("Error running Gerrit API operation " + url, e);
    }
  }

  @Override
  public <T> T post(String path, Object request, Type responseType)
      throws RepoException, ValidationException {
    HttpRequestFactory requestFactory = getHttpRequestFactory(getCredentials(uri.toString()));
    GenericUrl url = getUrl(path);
    try {
      return execute(responseType, requestFactory.buildPostRequest(
          url, new JsonHttpContent(JSON_FACTORY, request)));
    } catch (IOException e) {
      throw new RepoException("Error running Gerrit API operation " + url, e);
    }
  }

  @Override
  public <T> T put(String path, Object request, Type responseType)
      throws RepoException, ValidationException {
    HttpRequestFactory requestFactory = getHttpRequestFactory(getCredentials(uri.toString()));
    GenericUrl url = getUrl(path);
    try {
      return execute(responseType, requestFactory.buildPutRequest(
          url, new JsonHttpContent(JSON_FACTORY, request)));
    } catch (IOException e) {
      throw new RepoException("Error running Gerrit API operation " + url, e);
    }
  }

  public GenericUrl getUrl(String path) {
    Preconditions.checkArgument(path.startsWith("/"), path);
    return new GenericUrl(uri.resolve(uri.getPath() + path));
  }

  @SuppressWarnings("unchecked")
  public static <T> T execute(Type responseType, HttpRequest httpRequest)
      throws IOException, GerritApiException {
    HttpResponse response;
    try {
      response = httpRequest.execute();
    } catch (HttpResponseException e) {
      throw new GerritApiException(e.getStatusCode(), e.getContent(), e.getContent());
    }
    try {
      return (T) response.parseAs(responseType);
    } catch (IllegalArgumentException e) {
      throw new IllegalArgumentException(
          String.format("Cannot parse response as type %s.\n"
                  + "Request: %s\n"
                  + "Response:\n%s", responseType,
              httpRequest.getUrl(), response.parseAsString()), e);
    }
  }

  /**
   * TODO(malcon): Consolidate GitHub and this one in one class
   */
  private HttpRequestFactory getHttpRequestFactory(@Nullable UserPassword userPassword)
      throws RepoException, ValidationException {
    return httpTransport.createRequestFactory(
        request -> {
          request.setConnectTimeout((int) Duration.ofMinutes(1).toMillis());
          request.setReadTimeout((int) Duration.ofMinutes(1).toMillis());
          HttpHeaders httpHeaders = new HttpHeaders();
          if (userPassword != null) {
            httpHeaders.setBasicAuthentication(userPassword.getUsername(),
                                               userPassword.getPassword_BeCareful());
          }
          request.setHeaders(httpHeaders);
          request.setParser(new JsonObjectParser(JSON_FACTORY));
        });
  }

  /**
   * Credentials for API should be optional for any read operation (GET).
   *
   * TODO(malcon): Consolidate GitHub and this one in one class
   */
  @Nullable
  private UserPassword getCredentialsIfPresent(String url) throws RepoException {
    try {
      return getCredentials(url);
    } catch (ValidationException e) {
      return null;
    }
  }

  /**
   * Gets the credentials from git credential helper. First we try to get it for the api.github.com
   * host, just in case the user has an specific token for that url, otherwise we use the github.com
   * host one.
   *
   * TODO(malcon): Consolidate GitHub and this one in one class
   */
  private UserPassword getCredentials(String url) throws ValidationException, RepoException {
    try {
      return repo.credentialFill(url);
    } catch (ValidationException e) {
      throw new ValidationException(
          String.format("Cannot get credentials for host %s, from credentials helper", url), e);
    }
  }
}

File: src/Copybara.Core/Git/GerritApi/GerritApiTransportImpl.cs
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Json;

namespace Copybara.Git.GerritApi;

/// <summary>
/// Implementation of <see cref="IGerritApiTransport"/> that uses direct http calls. Port of
/// <c>com.google.copybara.git.gerritapi.GerritApiTransportImpl</c>.
/// </summary>
/// <remarks>
/// NOTE(port): the Java original uses google-http-client + gson. This port uses
/// <see cref="HttpClient"/> and <see cref="System.Text.Json"/>, and reuses
/// <see cref="GsonParserUtil"/> to strip Gerrit's <c>)]}'</c> XSSI no-execute prefix. Credentials
/// are obtained from the git credential helper via <see cref="GitRepository.CredentialFill"/>, and
/// applied as HTTP Basic auth, mirroring upstream.
/// </remarks>
public class GerritApiTransportImpl : IGerritApiTransport
{
    private static readonly TimeSpan Timeout = TimeSpan.FromMinutes(1);

    /// <summary>Serialization options mirroring gson: omit nulls, no indentation.</summary>
    private static readonly JsonSerializerOptions RequestOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly GitRepository _repo;
    private readonly Uri _uri;
    private readonly HttpClient _httpClient;

    public GerritApiTransportImpl(GitRepository repo, Uri uri, HttpClient httpClient)
    {
        _repo = repo;
        _uri = Preconditions.CheckNotNull(uri);
        _httpClient = Preconditions.CheckNotNull(httpClient);
    }

    public async Task<T?> GetAsync<T>(string path)
    {
        var userPassword = GetCredentialsIfPresent(_uri.ToString());
        var url = GetUrl(path);
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        return await ExecuteAsync<T>(request, userPassword, url).ConfigureAwait(false);
    }

    public async Task<T?> PostAsync<T>(string path, object request)
    {
        var userPassword = GetCredentials(_uri.ToString());
        var url = GetUrl(path);
        using var httpRequest = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = JsonContent(request),
        };
        return await ExecuteAsync<T>(httpRequest, userPassword, url).ConfigureAwait(false);
    }

    public async Task<T?> PutAsync<T>(string path, object request)
    {
        var userPassword = GetCredentials(_uri.ToString());
        var url = GetUrl(path);
        using var httpRequest = new HttpRequestMessage(HttpMethod.Put, url)
        {
            Content = JsonContent(request),
        };
        return await ExecuteAsync<T>(httpRequest, userPassword, url).ConfigureAwait(false);
    }

    public Uri GetUrl(string path)
    {
        Preconditions.CheckArgument(path.StartsWith('/'), path);
        return new Uri(_uri, _uri.AbsolutePath + path);
    }

    private static HttpContent JsonContent(object request)
    {
        string json = JsonSerializer.Serialize(request, request.GetType(), RequestOptions);
        return new StringContent(json, Encoding.UTF8, "application/json");
    }

    private async Task<T?> ExecuteAsync<T>(
        HttpRequestMessage request, GitCredential.UserPassword? userPassword, Uri url)
    {
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        if (userPassword != null)
        {
            var raw = $"{userPassword.GetUsername()}:{userPassword.GetPasswordBeCareful()}";
            var encoded = Convert.ToBase64String(Encoding.UTF8.GetBytes(raw));
            request.Headers.Authorization = new AuthenticationHeaderValue("Basic", encoded);
        }

        HttpResponseMessage response;
        try
        {
            using var cts = new CancellationTokenSource(Timeout);
            response = await _httpClient.SendAsync(request, cts.Token).ConfigureAwait(false);
        }
        catch (Exception e) when (e is HttpRequestException or OperationCanceledException)
        {
            throw new RepoException("Error running Gerrit API operation " + url, e);
        }

        using (response)
        {
            if (!response.IsSuccessStatusCode)
            {
                string content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                throw new GerritApiException(
                    (int)response.StatusCode, "Error calling gerrit", content, url.ToString());
            }

            try
            {
                return await GsonParserUtil
                    .ParseHttpResponseAsync<T>(response, stripNoExecutePrefix: true)
                    .ConfigureAwait(false);
            }
            catch (ArgumentException e)
            {
                throw new RepoException(
                    string.Format(
                        "Cannot parse response as type {0}.\nRequest: {1}\n", typeof(T), url),
                    e);
            }
        }
    }

    /// <summary>
    /// Credentials for API should be optional for any read operation (GET).
    /// </summary>
    private GitCredential.UserPassword? GetCredentialsIfPresent(string url)
    {
        try
        {
            return GetCredentials(url);
        }
        catch (ValidationException)
        {
            return null;
        }
    }

    /// <summary>Gets the credentials from the git credential helper.</summary>
    /// <exception cref="ValidationException"/>
    /// <exception cref="RepoException"/>
    private GitCredential.UserPassword GetCredentials(string url)
    {
        try
        {
            return _repo.CredentialFill(url);
        }
        catch (ValidationException e)
        {
            throw new ValidationException(
                $"Cannot get credentials for host {url}, from credentials helper", e);
        }
    }
}
